master/buildbot/util/gceclientservice.py

```diff
@@ -0,0 +1,233 @@
+
+        return url, req_kwargs
+
+    def zoneEndpoint(self, method=None):
+        base = "/compute/v1/projects/{0}/zones/{1}".format(
+            self.project, self.zone)
+        if method:
+            return "{0}/{1}".format(base, method)
+        else:
+            return base
+
+    def instanceEndpoint(self, method=None):
+        rel = "instances/{0}".format(self.instance)
+        if method:
+            rel = "{0}/{1}".format(rel, method)
+        return self.zoneEndpoint(rel)
+
+    def getInstanceState(self, fields=None):
+        params = {}
+        if fields is not None:
+            params = {'fields': fields}
+
+        return self.get(self.instanceEndpoint(), params=params)
+
+    def instanceStart(self):
+        return self.post(self.instanceEndpoint("start"))
+
+    def instanceStop(self):
+        return self.post(self.instanceEndpoint("stop"))
+
+    def createDisk(self, image=None, name=None, type=None):
+        return self.post(self.zoneEndpoint("disks"),
+            json={
+                "sourceImage": "projects/{0}/global/images/{1}".format(
+                    self.project, image),
+                "name": name,
+                "type": "projects/{0}/zones/{1}/diskTypes/{2}".format(
+                    self.project, self.zone, type)
+            }
+        )
+
+    def attachDisk(self, name=None, boot=True, index=0):
+        return self.post(self.instanceEndpoint("attachDisk"),
+            json={
+                "boot": boot,
+                "source": self.zoneEndpoint("disks/{0}".format(name)),
+                "deviceName": name,
+                "index": index
+            }
+        )
+
+    def detachDisk(self, name):
+        return self.post(self.instanceEndpoint("detachDisk"),
+            params={"deviceName": name}
+        )
+
+    def deleteDisk(self, disk_name):
+        return self.delete(
+            "{0}/disks/{1}".format(self.zoneEndpoint(), disk_name))
+
+    def setMetadata(self, fingerprint=None, items=[]):
+        return self.post(self.instanceEndpoint("setMetadata"),
+            json={
+                "fingerprint": fingerprint,
+                "items": items,
+                "kind": "compute#metadata"
+            }
+        )
+
+    @defer.inlineCallbacks
+    def waitInstanceState(self, desiredState):
+        state = yield self.getInstanceState(fields='status')
+        while state['status'] != desiredState:
+            time.sleep(0.5)
+            state = yield self.getInstanceState(fields='status')
+
+    @defer.inlineCallbacks
+    def waitForOperation(self, op):
+        op = yield op
+        while op['status'] != 'DONE':
+            time.sleep(0.5)
+            op = self.get(op['selfLink'])
+            op = yield op
+
+        if 'error' in op:
+            raise GCEError(op['error'])
+
+    # lets be nice to the auto completers, and don't generate that code
+    def get(self, ep, **kwargs):
```

Comment:
I think it would make sense to refactor HTTPClientService a little so that we don't need to duplicate this code here. What about renaming `_doRequest` to `_doHttpRequest` and having `get`, `put`, `post`, `delete` call `_doRequest` which would call `_doHttpRequest` in the default implementation? We would be able to override `_doRequest` in this class and call `_doHttpRequest` to actually perform the low-level request. `_doHttpRequest` may be changed to a better name indicating actual "low level" transport implementation, suggestions are welcome :-)